### DISCUSSION TITLE
macOS — Firebird 5.0.1.1469 — Too many open Files

### DISCUSSION THREAD

**Discussion by devnoname120** [FirebirdSQL/firebird#8345] [Q&A] [answered]
Posting this here so that other people who have that problem can find the solution because it really wasn't easy to find that the problem was with `csh` rather than sh, bash, or the system/kernel parameters.

I have the following error:
```shell
$ isql
SQL> CONNECT /Users/paul/STTDB3. FDB5;
Statement failed, SQLSTATE = 08001
I/0 error during "open" operation for file "/tmp/firebird/fb_tpc_0d000001ba2560600000000_8736"'
-Error while trying to open file
-Too many open files
```

And in JetBrains PyCharm and DataGrip I have this error:

```
DBMS: Firebird (no ver.)
Case sensitivity: plain=mixed, delimited=exact
[08001][335544344] I/O error during "open" operation for file "/tmp/firebird/fb_tpc_0f000001bae2560600000000_8729"; Error while trying to open file; Too many open files [SQLState:08001, ISC error code:335544344].
```

# My configuration:

- **Version**: `Firebird 5.0.1`
  - **Installation package**: `Firebird-5.0.1.1469-0-macos-arm64.pkg`
- **OS**: `macOS 14.7.1` (Sonoma) arm64
- **Kernel**: `Darwin 23.6.0`
- **Model**: MacBook Pro (16-inch, 2023)

**Comment by devnoname120** [comment, marked as answer]
# Solution ✅

How to fix the Firebird server service:
```shell
echo 'limit openfiles 65535' | sudo tee -a /etc/csh.login
```

How to fix `isql`:
```shell
echo 'ulimit -n 65535' | sudo tee -a /etc/profile
echo 'ulimit -n 65535' | sudo tee -a /etc/zprofile
```

-------------------

This should be enough. But if it still doesn't work you can attempt to run these in addition:
```shell
sudo launchctl limit maxfiles 65535 unlimited
sudo sysctl -w kern.maxfiles=245760
sudo sysctl -w kern.maxfilesperproc=122880
```

If these 3 last commands fixed it you will need to run them at boot to persist them. You can create two launch daemon files for that:

`/Library/LaunchDaemons/limit.maxfiles.plist`:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>limit.maxfiles</string>
    <key>ProgramArguments</key>
    <array>
        <string>/bin/launchctl</string>
        <string>limit</string>
        <string>maxfiles</string>
        <string>65535</string>
        <string>unlimited</string>
    </array>
    <key>RunAtLoad</key>
    <true/>
</dict>
</plist>
```

and

`/Library/LaunchDaemons/kern.maxfiles.plist`:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>kern.maxfiles</string>
    <key>ProgramArguments</key>
    <array>
        <string>/usr/sbin/sysctl</string>
        <string>-w</string>
        <string>kern.maxfiles=245760</string>
        <string>kern.maxfilesperproc=122880</string>
    </array>
    <key>RunAtLoad</key>
    <true/>
</dict>
</plist>
```

**Comment by devnoname120** [comment]
@mrotteveel Is it not considered a bug? I only have a single (admittedly big) database. I don't know why Firebird attempts to open so many files (caching maybe?) but shouldn't it restrict itself to the maximum allowed in the environment?

**Reply by aafemt** [reply]
Firebird opens as many files as it needs. OS limits is not Firebird issue.

**Reply by devnoname120** [reply]
@aafemt Why does it need to open > 256 files when I'm just opening a single database file?

**Reply by aafemt** [reply]
The file in starting post is opened for TIP page cache, for example. Other files are opened to store monitoring tables data, interprocess communication, lock table, events, etc. You can list them all using system utilities.

**Reply by devnoname120** [reply]
@aafemt Thank you for the explanations. IMO it's an issue that Firebird tries to open more files than the default allows. I'd suggest to either explicitly ask the user to increase this limit, or do it automatically. What do you think?

**Reply by mrotteveel** [reply]
> @mrotteveel Is it not considered a bug? I only have a single (admittedly big) database. I don't know why Firebird attempts to open so many files (caching maybe?) but shouldn't it restrict itself to the maximum allowed in the environment?

It doesn't read as a bug report or a request to the maintainers to fix/change something, but as a solution to a problem you have on macOS. As such, I thought it was more suitable to migrate it to a Discussion.

**Reply by devnoname120** [reply]
@mrotteveel Good point!

**Reply by mrotteveel** [reply]
In any case, if you think something should be changed in Firebird, create a new issue, and be explicit and specific about what you want to have addressed.

**Comment by AlexPeshkoff** [comment]
On 12/9/24 15:07, Paul wrote:
>
> @mrotteveel <https://github.com/mrotteveel> Good point!
>

Will be interesting to know - what particular files are opened?
